internal/net/http/transport/roundtrip.go

```diff
@@ -41,20 +42,37 @@ func NewExpBackoff(opts ...Option) http.RoundTripper {
 	return e
 }
 
+// RoundTrip round trip the HTTP request and return the response.
+// If backoff is not set, the default roundTrip implementation will be used.
+// It round trip the request and returns the response, and return any error occurred.
+// It returns errors.ErrTransportRetryable to indicate if the request is consider as retryable.
 func (e *ert) RoundTrip(req *http.Request) (res *http.Response, err error) {
 	if e.bo == nil {
 		return e.roundTrip(req)
 	}
-	var r interface{}
-	r, err = e.bo.Do(req.Context(), func() (interface{}, error) {
-		return e.roundTrip(req)
+
+	var ferr error
+	r, err := e.bo.Do(req.Context(), func() (res interface{}, err error) {
```

Comment:
could you please fix to named return.

```
var r interface{}
r, err = e.bo.Do
```

Reply:
do you mean this?
```suggestion
	res, err = e.bo.Do(req.Context(), func() (res interface{}, err error) {
```

Reply:
yes!

Reply:
no.
```
func (e *ert) RoundTrip(req *http.Request) (res *http.Response, err error) {
	if e.bo == nil {
		return e.roundTrip(req)
	}

	var ferr error
	var r interface{}
	r, err = e.bo.Do(req.Context(), func() (res interface{}, err error) {
		....
	})
	if err != nil {
		return nil, err
	}
	if ferr != nil {
		return nil, ferr
	}
 
	var ok bool
	res, ok = r.(*http.Response)
	if !ok {
		return nil, errors.ErrInvalidTypeConversion(r, res)
	}
	return res, nil
}
```

Reply:
I think we cannot do it as `bo.Do` returns `interface{}`

Reply:
commented below

Reply:
hmm, also I think we can reduce reflection 
stop using named return in Do function & directry insert to outer named return res variable

Reply:
I also thought so. Personally I don't like named return value as I think it is confusing XDD

Reply:
but it reduce memory allocation time and It's memory efficient.
I think we should use named return by default, but we should be careful about naming for each return values

Reply:
yes :( we need to be careful when we use it, especially in multi-layer functions

Reply:
if I change the implementation below, the return response will becomes nil, do you have any idea?
```golang
 49 func (e *ert) RoundTrip(req *http.Request) (res *http.Response, err error) {
 50     if e.bo == nil {
 51         return e.roundTrip(req)
 52     }
 53
 54     var ferr error
 55     _, err = e.bo.Do(req.Context(), func() (interface{}, error) {
 56         res, err = e.roundTrip(req)
 57         if err != nil {
 58             // if the error is retryable, return the error and let backoff to retry.
 59             if errors.Is(err, errors.ErrTransportRetryable) {
 60                 return nil, err
 61             }
 62             // if the error is not retryable, return nil error to terminate the backoff execution
 63             ferr = err
 64             return nil, nil
 65         }
 66         return res, nil
 67     })
 68     if err != nil {
 69         return nil, err
 70     }
 71     if ferr != nil {
 72         return nil, ferr
 73     }
 74
 75     /*
 76         res, ok := r.(*http.Response)
 77         if !ok {
 78             return nil, errors.ErrInvalidTypeConversion(r, res)
 79         }
 80     */
 81
 82     return res, nil
 83 }
```

Reply:
sorry it made some mistake in my test code, I have fixed it.